Fix behavior of nested click event on label element with checkbox

Ahmad-S792 opened the pull request:
#### 4ee89ea3f12cb36cc05e36f80c77c73553e62d28
<pre>
Fix behavior of nested click event on label element with checkbox

Fix behavior of nested click event on label element with checkbox
<a href="https://bugs.webkit.org/show_bug.cgi?id=249789">https://bugs.webkit.org/show_bug.cgi?id=249789</a>

Reviewed by Ryosuke Niwa.

This patch is to align WebKit with Blink / Chromium and Gecko / Firefox.

Merge - <a href="https://src.chromium.org/viewvc/blink?view=revision&amp">https://src.chromium.org/viewvc/blink?view=revision&amp</a>;revision=183243

This patch removes static variable in HTMLLabelElement::defaultEventHandler, which interferes with handling of nested event on another label element. This is to promote this variable to class variable in the header so this interference can be avoided.

* Source/WebCore/html/HTMLLabelElement.cpp:
(HTMLLabelElement::defaultEventHandler): removed static variable and rename usage with SetForScope
* Source/WebCore/html/HTMLLabelElement.h: Add bool &apos;m_processingClick&apos; with false
* LayoutTests/fast/form/checkbox-nested-click-event-on-label.html: Add Test Case
* LayoutTests/fast/form/checkbox-nested-click-event-on-label-expected.txt: Add Test Case Expectation

Canonical link: <a href="https://commits.webkit.org/258287@main">https://commits.webkit.org/258287@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/b86c523ff71606b3654c642526d35766c098edf7

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/101435 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/10593 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/34489 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/110705 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/170963 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/105415 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/11544 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/1446 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/93823 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/108522 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/107217 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/8779 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/92019 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/35302 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/90678 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/23427 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/78302 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/4195 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/24938 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/4257 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/1365 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/10344 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/44426 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5700 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/6020 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/html/HTMLLabelElement.cpp:
```diff
@@ -151,15 +151,15 @@ void HTMLLabelElement::defaultEventHandler(Event& event)
         }
 
-        processingClick = true;
+        m_processingClick = true;
```
Can we use SetForScope?

Ahmad-S792 replied:
I just updated it, can you have another look whether I used SetForScope properly, first time using it. :-)

Ahmad-S792 replied:
Done.

rniwa commented on Source/WebCore/html/HTMLLabelElement.cpp:
```diff
@@ -151,15 +152,15 @@ void HTMLLabelElement::defaultEventHandler(Event& event)
             control->focus({ { }, { }, { }, FocusTrigger::Click, { } });
 
-        processingClick = false;
+        SetForScope processingClick(m_processingClick, false);
```
Oh you don't need this second SetForScope. A neat thing about SetForScope is that it would restore the original value in its destructor.

Ahmad-S792 replied:
Also should I add value of { false } in header (HTMLLabelElement.h) and then remove global variable in HTMLLabelElement.cpp?

rniwa replied:
Sure, initializing in the header file makes sense to me.

Ahmad-S792 replied:
Updated commit message and also did all changes. Now let's see.

rniwa reviewed (approved):
r=me

Ahmad-S792 committed:
Fix behavior of nested click event on label element with checkbox

Fix behavior of nested click event on label element with checkbox
https://bugs.webkit.org/show_bug.cgi?id=249789

Reviewed by Ryosuke Niwa.

This patch is to align WebKit with Blink / Chromium and Gecko / Firefox.

Merge - https://src.chromium.org/viewvc/blink?view=revision&revision=183243

This patch removes static variable in HTMLLabelElement::defaultEventHandler, which interferes with handling of nested event on another label element. This is to promote this variable to class variable in the header so this interference can be avoided.

* Source/WebCore/html/HTMLLabelElement.cpp:
(HTMLLabelElement::defaultEventHandler): removed static variable and rename usage with SetForScope
* Source/WebCore/html/HTMLLabelElement.h: Add bool 'm_processingClick' with false
* LayoutTests/fast/form/checkbox-nested-click-event-on-label.html: Add Test Case
* LayoutTests/fast/form/checkbox-nested-click-event-on-label-expected.txt: Add Test Case Expectation

Canonical link: https://commits.webkit.org/258287@main
